[Integrations][System] - Fixed user.target.domain extraction for nested domains

m-hoskin-unisa opened the pull request:
## Type of change
- Bug

## What does this PR do?

The security pipeline tries to extract the value of the user's domain for certain kinds of events (e.g., user added to a domain group) and stores it into user.target.domain. The current script is hard coded to split the member name on comma, and then extract a specific item in the array, which will only be valid for domains that are not nested under more than domain (e.g., AD_domain.company_domain.com.country_code will extract the company_domain rather than the AD_domain.

This PR updates the script to instead use a for loop to iterate over the array and match against the item that starts with "DC=" and use that for user.target.domain.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [ ] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Supersedes #123
-->
-

## Screenshots

<!-- Optional
Add here screenshots presenting:
- Kibana UI forms presenting configuration options exposed by the integration
- dashboards with collected metrics or logs
-->

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > The PR is not allowed to run in the CI yet
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-7629/detail/PR-7629/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-7629/detail/PR-7629/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-7629/detail/PR-7629/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-7629/detail/PR-7629/1//artifacts)   [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2023-09-01T05:50:04.931Z&rangeTo=2023-09-01T06:10:04.931Z&transactionName=BUILD+Ingest-manager%2Fintegrations%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=3136d15639b16cf9518cb6b1f6f54866&transactionId=2a178ac5ca137686)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-09-01T06:00:04.931+0000


* Duration: 3 min 7 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![2](https://img.shields.io/badge/2%20-red)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-7629/detail/PR-7629/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Load a resource file from a library`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-7629/runs/1/steps/81/log/?start=0">here</a></li>
<li>Description: <code>approval-list/elastic/integrations.yml</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-7629/runs/1/steps/89/log/?start=0">here</a></li>
<li>Description: <code>githubApiCall: The REST API call https://api.github.com/orgs/elastic/members/m-hoskin-unisa return the message : java.lang.Exception: httpRequest: Failure connecting to the service https://api.github.com/orgs/elastic/members/m-hoskin-unisa : httpRequest: Failure connecting to the service https://api.github.com/orgs/elastic/members/m-hoskin-unisa : Code: 404Error: {"message":"User does not exist or is not a member of the organization","documentation_url":"https://docs.github.com/rest/orgs/members#check-organization-membership-for-a-user"}</code></l1>
</ul>
  
  </p>
  </details>



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

bhapas commented on packages/system/data_stream/security/elasticsearch/ingest_pipeline/standard.yml:
```diff
@@ -2584,8 +2584,13 @@ processors:
-            ctx.user.target.put("domain", domain);
+            for (int i = 0; i < memberNameParts.length; i++) {
+              def domainString = memberNameParts[i];
+              if ((domainString.contains("DC=") || domainString.contains("dc=")) && ctx.user.target.domain == null) {
```
```suggestion
              if ((domainString.contains("DC=") || domainString.contains("dc=")) && ctx.user?.target?.domain == null) {
```

bhapas reviewed (changes requested):
@m-hoskin-unisa 

Thank you for the PR. There are a few things to check off before we can take this PR in.

- [x] Please sign CLA - [Contributor Agreement](https://www.elastic.co/contributor-agreement)
- [ ] Please add some pipeline tests for this change
- [ ] Please change the version in `manifest.yml` too

m-hoskin-unisa commented:
@bhapas thanks for the suggested improvement, I've added it to the PR. I updated the metadata file and have signed the contributor agreement. I'm not too familiar with sending a PR to another's project so I'm not sure how to add tests for this request.

bhapas commented:
> @bhapas thanks for the suggested improvement, I've added it to the PR. I updated the metadata file and have signed the contributor agreement. I'm not too familiar with sending a PR to another's project so I'm not sure how to add tests for this request.

@m-hoskin-unisa Sorry for a late reply. Not sure how this got missed. You can add pipeline tests and take reference from here https://github.com/elastic/integrations/tree/main/packages/system/data_stream/security/_dev/test/pipeline

efd6 commented on packages/system/data_stream/security/elasticsearch/ingest_pipeline/standard.yml:
```diff
@@ -2584,8 +2584,13 @@ processors:
-            ctx.user.target.put("domain", domain);
+            for (int i = 0; i < memberNameParts.length; i++) {
+              def domainString = memberNameParts[i];
+              if ((domainString.contains("DC=") || domainString.contains("dc=")) && ctx.user?.target?.domain == null) {
```
ISTM this is doing more work than is necessary. I would suggest adding a helper above
```
String trimPrefix(String s, String prefix) {
    if (s.startsWith(prefix)) {
        return s.substring(prefix.length());
    }
    return s;
}
```
and then the code here becomes
```
if (ctx.user?.target?.domain != null) {
    ctx.user.target.domain = trimPrefix(trimPrefix(ctx.user.target.domain, "DC="), "dc=);
}
```

efd6 reviewed (commented):
Please add a test case for this.
